### ISSUE TITLE
Finalizing of interrupted background migrations not possible

### ISSUE THREAD

**Issue by dirkhillbrecht** [sameersbn/docker-gitlab#2460]
Hi everyone, I have the infamous background migration problem on updating from 14.0.6 to 14.4.1:

    /home/git/gitlab/lib/gitlab/database/migration_helpers.rb:1109:in `ensure_batched_background_migration_is_finished'
    [...]
    /usr/local/bin/bundle:23:in `load'
    /usr/local/bin/bundle:23:in `<main>'
    
    Caused by:
    Expected batched background migration for the given configuration to be marked as 'finished', but it is 'active':    
    {:job_class_name=>"CopyColumnUsingBackgroundMigrationJob", :table_name=>"events", :column_name=>"id", :job_arguments=>[["id"], ["id_convert_to_bigint"]]}
    
    Finalize it manualy by running
    
        sudo gitlab-rake gitlab:background_migrations:finalize[CopyColumnUsingBackgroundMigrationJob,events,id,'[["id"]\, ["id_convert_to_bigint"]]']

After poking a bit backward and forward I transcribed this in a script:

    #!/bin/bash
    docker run -it --rm --name gitlab \
    --link ... -- publish ... --env ... --volume ... \
    sameersbn/gitlab:14.4.1 app:rake "gitlab:background_migrations:finalize[CopyColumnUsingBackgroundMigrationJob,events,id,'[[\"id\"]\\,[\"id_convert_to_bigint\"]]']"

This seems to me to pass all characters through into the rake task. Unfortunately, I get errors:

    [...]
    Running raketask gitlab:background_migrations:finalize[CopyColumnUsingBackgroundMigrationJob,events,id,'[["id"]\,["id_convert_to_bigint"]]']...
    Missing Rails.application.secrets.openid_connect_signing_key for production environment. The secret will be generated and stored in config/secrets.yml.
    WARNING: This installation of GitLab uses a deprecated syntax for 'config/database.yml'. The support for this syntax will be removed in 15.0. More information can be found here: https://gitlab.com/gitlab-org/gitlab/-/issues/338182
    rake aborted!
    JSON::ParserError: unexpected character () at line 1, column 1 [parse.c:714]
    /home/git/gitlab/lib/gitlab/json.rb:87:in `rescue in adapter_load'
    /home/git/gitlab/lib/gitlab/json.rb:82:in `adapter_load'
    /home/git/gitlab/lib/gitlab/json.rb:25:in `parse'
    /home/git/gitlab/lib/tasks/gitlab/background_migrations.rake:17:in `block (3 levels) in <top (required)>'
    /home/git/gitlab/vendor/bundle/ruby/2.7.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
    /usr/local/bin/bundle:23:in `load'
    /usr/local/bin/bundle:23:in `<main>'
    
    Caused by:
    Oj::ParseError: unexpected character () at line 1, column 1 [parse.c:714]
    /home/git/gitlab/lib/gitlab/json.rb:85:in `load'
    /home/git/gitlab/lib/gitlab/json.rb:85:in `adapter_load'
    /home/git/gitlab/lib/gitlab/json.rb:25:in `parse'
    /home/git/gitlab/lib/tasks/gitlab/background_migrations.rake:17:in `block (3 levels) in <top (required)>'
    /home/git/gitlab/vendor/bundle/ruby/2.7.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
    /usr/local/bin/bundle:23:in `load'
    /usr/local/bin/bundle:23:in `<main>'
    Tasks: TOP => gitlab:background_migrations:finalize

And now I am really stuck. I have no idea whether this is a bug in the containered software, a call error on my behalf or if the database is broken. Can anyone enlighten me what is going on here, if there is any way to fix it or where I can find help? Regards, Dirk

**Comment by dirkhillbrecht** [comment]
Hi, thanks for the quick reply! Unfortunately, it does not work. Using

    '[[\"id\"]\,[\"id_convert_to_bigint\"]]'

yields exactly the same error message.

Note that there is another difference between the gitlab docs and what I made of them: They add a space after the comma:

    '[[\"id\"]\, [\"id_convert_to_bigint\"]]'

However, I was absolutely not able to conserve this space through the call stack. The rake task always only gets the String up to the comma as parameter if I put a space, an escaped space, a double-escaped space there.

    Running raketask gitlab:background_migrations:finalize[CopyColumnUsingBackgroundMigrationJob,events,id,'[["id"]\, ["id_convert_to_bigint"]]']...
    rake aborted!
    Don't know how to build task
    'gitlab:background_migrations:finalize[CopyColumnUsingBackgroundMigrationJob,events,id,'[["id"]\,' (See the list of available tasks with `rake --tasks`)
    /home/git/gitlab/vendor/bundle/ruby/2.7.0/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
    /usr/local/bin/bundle:23:in `load'
    /usr/local/bin/bundle:23:in `<main>'

I do not know if this space is important, I only know that this is a difference. Otherwise, I am still stuck with the problem, even though I eliminated a number of other mistakes I made during the upgrade process. Those I ironed out by starting from scratch with a clean backup of the gitlab instance to be updated. Quite a number of other problems solved, this persists. Regards!

**Comment by kkimurak** [comment]
Hi, thank you for prompt response.  

I have confirmed that same error with same command. After some trial and error I have executed following. It successfully done. **No need to double-quote rake task name, need to escape `"`.**  

```sh
$ docker-compose run --rm gitlab app:rake gitlab:background_migrations:finalize[CopyColumnUsingsBackgroundMigrations,events,id,'[[\"id\"]\,[\"id_convert_to_bigint\"]]']
```

P.S. The rake task name shown in your log is [just a first argument for `app:rake`](https://github.com/sameersbn/docker-gitlab/blob/7665a499af62ddf2f67e49548c6bd6f795b80d84/assets/runtime/functions#L2008). White space is field separator so I think all of arguments for [`app:rake`](https://github.com/sameersbn/docker-gitlab/blob/7665a499af62ddf2f67e49548c6bd6f795b80d84/entrypoint.sh#L37-L40) have been [bypassed to `bundle exec rake`](https://github.com/sameersbn/docker-gitlab/blob/7665a499af62ddf2f67e49548c6bd6f795b80d84/assets/runtime/functions#L2009) correctly even the log only shows task name before white space.

**Comment by dirkhillbrecht** [comment]
Hi, thanks for your reply.

Fortunately (or, in the context of this issue, unfortunately) I was able to solve the problem by performing the whole upgrade process again - and this time correctly. It turned out that I had interrupted multiple upgrade steps unintentionally. That trashed the database scheme at multiple occasions. I restarted the upgrade from scratch from a healthy backup and performed step by step waiting until everything had finished. This went through flawlessly and now I have a working and current version of gitlab on our server.

So, I cannot try your suggested way any more. But perhaps it helps others in a similar situation who cannot use a backup to start over.

Kind regards, Dirk
